Add documentation for Maintenance Mode and Health Checks in Front-Commerce (#917)
front-commerce/developers.front-commerce.com: docs/02-guides/maintenance-mode/automatic-detection-with-service-health-checks.mdx

back up, the maintenance mode API will automatically remove the store from
maintenance mode.

Additionally, you can configure the scheduling of the health checks using a Cron
pattern. This allows you to control how often the health checks are run and how
quickly your store is put in maintenance mode if a service is down.

:::info

When Maintenance Mode has been activated manually via
[the maintenance mode API](./entering-and-exiting-maintenance-mode#the-maintenance-mode-api),
the health checks will not remove the store from maintenance mode even after a
service has recovered. To remove the store from maintenance mode, you must
[manually deactivate](./entering-and-exiting-maintenance-mode#deactivating-the-maintenance-mode)
it.

:::

### Adding a Service Health Check

To add a health check, you need to define a function that checks the health of
your service and returns:

- `true`: when the service is `up`
- `false`: when the service is `down`
- `thrown Error`: assumes the service is `down`

This function is then passed to `addHealthCheckService` in the `onServerInit`
hook in your extension definition.

Here is an example:

> ```ts title="extensions/health-check.ts"

Reviewer: **thought:** Shouldn't this path be in an extension, rather than in the extensions' "root" directory?

```suggestion
```ts title="extensions/acme/health-check.ts"
```

Reviewer: I do believe we should encourage integrator to organise their extension in separated folders, even if those are 1 file big, but I guess you're right.

import { defineExtension } from "@front-commerce/core";

export default defineExtension({
  name: "health-check",
  meta: import.meta,
  unstable_lifecycleHooks: {
    onServerInit: async (services, request, config) => {
      // highlight-start
      services.MaintenanceMode.addHealthCheckService("acme", async () => {
        const response = await fetch("https://acme.com/api/health-check");
        return response.status !== 200;
      });
      // highlight-end
    },
  },
});
```

### Configuring the global Cron interval

By default, the health checks are run every 10 seconds. You can configure a

> custom interval by providing a Cron pattern in your `front-commerce.config.ts`

Reviewer: **thought:** It might be helpful for integrators that are not familiar with Cron syntax to provide a link to something explaining it. I.E. [https://en.wikipedia.org/wiki/Cron](https://en.wikipedia.org/wiki/Cron)

---
title: Automatic detection with service Health Checks
description:

>   Learn how to implement health checks to detect down services and automatically
>   place your site in maintenance mode.

Reviewer 2: **suggestion:** 
```suggestion
  Learn how to implement health checks to place your site in maintenance mode
  as soon as a remote services is unavailable.
```

sidebar_position: 2
---

<SinceVersion tag="3.6" />

# Health Checks

Health checks allow you to monitor the availability of your services and
automatically put your store in maintenance mode if a service is down. This
ensures that your users are not affected by a service outage.

With health checks, you can define a set of services that you want to monitor

> and provide a function to check the health of each service. The maintenance mode
> API will periodically run these health check functions and if a service is down,
> it will automatically put the store in maintenance mode. Once the service is
> back up, the maintenance mode API will automatically remove the store from
> maintenance mode.

Reviewer 2: **nitpick:** at this level, I don't think mentioning the "Maintenance mode API" is relevant.

**suggestion:** use "Front-Commerce", which is more generic and (thus) futureproof

```suggestion
and provide a function to check the health of each service. Front-Commerce
will periodically run these health check functions and if a service is down,
it will automatically put the store in maintenance mode. Once the service is
back up, the Front-Commerce will automatically remove the store from
maintenance mode.
```
